### PR TITLE
Apache rollback

### LINKED ISSUE #1394: Audit Apache plugin for guaranteed reversion
We've had some apache tickets (eg #998 #1354 #1243) which imply that the plugin's revert-in-case-of-error mechanisms are imperfect.  Those completely block the release of the Apache plugin.  We need to perform an audit to ensure that the all of the Apache plugin's file writing operations, module enablement/disablement, and config editing are properly wrapped in proper reversion logic and try-except blocks that ensure reversion is attempted after any failure.

### PR DESCRIPTION
**Pull request by bmw**
Fixes #1394. With this PR, I'm confident in our ability to rollback changes made by the Apache plugin.

Related issues are #808 and #1456 but neither are showstoppers for launch.

### CODE REVIEW

File: letsencrypt/client.py [hunk 1]
```diff
@@ -117,6 +120,12 @@ class Client(object):
 
     """
 
+    # Message to show if enabling a redirect fails, but recovery is successful
+    _SUCCESSFUL_REDIRECT_RECOVERY_MSG = (
+        "We were unable to setup a redirect for your server, however, we "
+        "successfully installed your certificate. If you'd like to revert "
+        "these changes as well, run 'letsencrypt rollback'.")
```
Comment (pde) [inline letsencrypt/client.py]: This is confusing language.  Does the user know what other reversion is referred to by the words "as well"?  Do they need to know?  I think probably we should just strike "~~If you'd like to revert these changes as well, run letsencrypt rollback~~", but if we want to keep it, find a narrative of when and why they would want to run rollback now.

"setup" -> "set up"?

File: letsencrypt/client.py [hunk 2]
```diff
@@ -331,25 +340,14 @@ def deploy_certificate(self, domains, privkey_path,
             self.installer.save("Deployed Let's Encrypt Certificate")
 
         # sites may have been enabled / final cleanup
-        with error_handler.ErrorHandler(self._rollback_and_restart):
+        restart_error_handler = error_handler.ErrorHandler(functools.partial(
+            self._rollback_and_restart,
+            "We were unable to install your certificate, however, we "
+            "successfully rolled back your server to its previous "
```
Comment (pde) [inline letsencrypt/client.py]: "previous" -> "prior"?

File: letsencrypt/client.py [hunk 3]
```diff
@@ -1,10 +1,12 @@
 """Let's Encrypt client API."""
+import functools
```
Comment (pde) [inline letsencrypt/client.py]: Maybe `from functools import partial`?  Personally I find the fun of functional programming is gone if it's extremely verbose :)

**Comment by joohoi** [comment]
LGTM codewise. However I'd refactor the return messages (closer) to layman's terms. 

Eg. "...we successfully rolled back your server to its previous configuration" would make John Doe happier and more confident if it'd say for example: "...we successfully restored your web server back to its previous configuration"

File: letsencrypt/client.py [hunk 4]
```diff
@@ -331,25 +340,14 @@ def deploy_certificate(self, domains, privkey_path,
             self.installer.save("Deployed Let's Encrypt Certificate")
 
         # sites may have been enabled / final cleanup
-        with error_handler.ErrorHandler(self._rollback_and_restart):
+        restart_error_handler = error_handler.ErrorHandler(functools.partial(
+            self._rollback_and_restart,
+            "We were unable to install your certificate, however, we "
```
Comment (joohoi) [inline letsencrypt/client.py]: Could we make a class variable of this message as well, like _SUCCESSFUL_REDIRECT_RECOVERY_MSG. That'd be a consistent way of managing the different return messages.

File: letsencrypt/client.py [hunk 5]
```diff
@@ -395,7 +396,43 @@ def redirect_to_ssl(self, domains):
...
+        """
+        logger.critical("Rolling back to previous server configuration...")
+        reporter = zope.component.getUtility(interfaces.IReporter)
+        try:
+            self.installer.rollback_checkpoints()
             self.installer.restart()
+        except:
+            # TODO: suggest letshelp-letsencypt here
+            reporter.add_message(
+                "An error occured and we failed to rollback your config and "
```
Comment (joohoi) [inline letsencrypt/client.py]: Could we make a class variable of this message as well, like _SUCCESSFUL_REDIRECT_RECOVERY_MSG. That'd be a consistent way of managing the different return messages.

**Comment by joohoi** [comment]
LGTM

**Comment by pde** [comment]
Great!
